### PR TITLE
8264409: AArch64: generate better code for Vector API allTrue

### PR DESCRIPTION
**Pull request by nsjian**
In Vector API NEON implementation, we use a vector register to represent vector mask, where an element value of -1 is a true mask and an element value of 0 is a false mask. The allTrue() api is used to check whether all the elements of current mask are set.

Currently, the AArch64 NEON allTrue implementation looks like:

  andr  $tmp, T16B $src1, $src2\t# src2 is maskAllTrue
  notr  $tmp, T16B, $tmp
  addv  $tmp, T16B, $tmp
  umov  $dst, $tmp, B, 0
  cmp   $dst, 0
  cset  $dst

where $src2 is a preset all true (-1) constant. We can optimize it to the code sequence like below, to check whether all bits are set:

  uminv $tmp, T16B, $src1
  umov  $dst, $tmp, B, 0
  cmp   $dst, 0xff
  cset  $dst

With this codegen improvement, we can see about 8%~70% performance uplift on different machines for Alibaba's Vector API bigdata benchmarks [1][2].

Tested with tier1 and vector api jtreg tests.

[1] https://github.com/openjdk/panama-vector/blob/vectorIntrinsics/test/jdk/jdk/incubator/vector/benchmark/src/main/java/benchmark/bigdata/BooleanArrayCheck.java#L61
[2] https://github.com/openjdk/panama-vector/blob/vectorIntrinsics/test/jdk/jdk/incubator/vector/benchmark/src/main/java/benchmark/bigdata/ValueRangeCheckAndCastL2I.java#L93

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8264409](https://bugs.openjdk.java.net/browse/JDK-8264409): AArch64: generate better code for Vector API allTrue

### Reviewers
 * [Andrew Dinn](https://openjdk.java.net/census#adinn) (@adinn - **Reviewer**)
 * [Pengfei Li](https://openjdk.java.net/census#pli) (@pfustc - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/3302/head:pull/3302` \
`$ git checkout pull/3302`

Update a local copy of the PR: \
`$ git checkout pull/3302` \
`$ git pull https://git.openjdk.java.net/jdk pull/3302/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 3302`

View PR using the GUI difftool: \
`$ git pr show -t 3302`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/3302.diff">https://git.openjdk.java.net/jdk/pull/3302.diff</a>

</details>

### CODE REVIEW

**Review by adinn — APPROVED** [review]
It's a clever trick to use uminv for this specific case.
The patch looks good to me.

File: src/hotspot/cpu/aarch64/aarch64_neon.ad [hunk 1]
```diff
@@ -3564,19 +3566,15 @@ instruct alltrue_in_mask8B(iRegINoSp dst, vecD src1, vecD src2, vecD tmp, rFlags
...
   ins_cost(INSN_COST);
   effect(TEMP tmp, KILL cr);
-  format %{ "andr  $tmp, T8B, $src1, $src2\t# src2 is maskAllTrue\n\t"
-            "notr  $tmp, T8B, $tmp\n\t"
-            "addv  $tmp, T8B, $tmp\n\t"
+  format %{ "uminv $tmp, T8B, $src1\n\t"
             "umov  $dst, $tmp, B, 0\n\t"
-            "cmp   $dst, 0\n\t"
-            "cset  $dst" %}
+            "cmp   $dst, 0xff\n\t"
```
Comment (pfustc) [inline src/hotspot/cpu/aarch64/aarch64_neon.ad:3571]: I think we should write "#0xff" here. But it looks that all other immediates in format field of aarch64_neon.ad lose the number sign as well.

**Review by pfustc — APPROVED** [review]
Overall looks good to me. (not a reviewer)

File: src/hotspot/cpu/aarch64/aarch64_neon.ad (hunk 1 above)
Comment (nsjian) [reply src/hotspot/cpu/aarch64/aarch64_neon.ad:3571]: Thanks for the review, but I think both are ok.

**Comment by nsjian** [comment]
> It's a clever trick to use uminv for this specific case.
> The patch looks good to me.

Thank you for the review! @adinn

**Pull request closed without merging**